### DISCUSSION TITLE
Measure the distance between two named anchor points

### DISCUSSION THREAD

**Discussion by taylorh140** [cetz-package/cetz#561] [Q&A]
I'm trying to mimic obsidian/other infinite canvas lines.

I think they are bezier curves but I'm having problems with two properties:

```typst
    bezier( 
           ptA,
           ptB,
           (to:ptA,rel:ctrlpointfrom.at(edge.fromSide)), 
           (to:ptB,rel:ctrlpointto.at(edge.toSide) ),
           stroke:ec, 
           name:edge.id
        )
 ```
 
 with this i have an adjust ment where i add a control point to the left or right by a constant amount of units and it can look alright:
 
![image](https://github.com/cetz-package/cetz/assets/10996543/178bf7d2-f81c-4ede-a45b-8267c09c6882)

but it needs to be fairly large to work nicely and for short elements they tend to hop accros the endponts

for example these are suppose to be pointing down:
![image](https://github.com/cetz-package/cetz/assets/10996543/7c94cdc9-6d97-4c7d-ae05-a02bcf1fe263)

like so:
![image](https://github.com/cetz-package/cetz/assets/10996543/1d9e84a1-7798-4376-99bb-d114667236ba)

I think maybe i could adjust the setpoints by how far the endpoints are but i dont know how to measure them in cetz.

**Comment by johannes-wolf** [comment]
To get vectors from coordinates, use something like `let (_, v) = coordinate.resolve(ctx, ...)` insides a `get-ctx` call. I will add an example when I am at home.

**Reply by taylorh140** [reply]
Got it!, how about like this:

```typst
#let distance(a,b) = {
  let (x1,y1,z1) = a
  let (x2,y2,z2) = b
  return calc.sqrt((x1 - x2)*(x1 - x2) + (y1 - y2)*(y1 - y2))
} 

#cetz.canvas({
  import cetz.draw: *
  get-ctx( ctx => {
      // ptA and ptB defined here somwhere
      let (ctx, a, b) = cetz.coordinate.resolve(ctx, ptA, ptB)
      let dist = distance(a,b)
    })
})
```

**Reply by fenjalien** [reply]
Yes that would work a two points though:
- Theres a vector distance function `cetz.vector.dist` that is used internally
- You could use a function coordinate depending on your use case.

**Reply by johannes-wolf** [reply]
Good idea! I totally forgot about function coordinates.
With function coordinates, it could look like this:
```typst
#import "@preview/cetz:0.2.2"

#cetz.canvas(length: 1cm, {
  import cetz.draw: *

  let connect(a, b, horizontal: true) = {
    let ctrl-pos(a, b) = {
      let (dx, dy, ..) = cetz.vector.sub(b, a)
      cetz.vector.add(a, if horizontal {
        (dx / 3,0,0)
      } else {
        (0, dy / 3,0)
      })
    }
    
    bezier(a, b, (ctrl-pos, a, b), (ctrl-pos, b, a))
  }

  connect((0,0), (2, 2))
  connect((0,0), (2, 1))
  connect((0,0), (2, 0))
  connect((0,0), (2, -1))
  connect((0,0), (2, -2))

  set-origin((0, 2))
  
  connect((0,0), (2, 2), horizontal: false)
  connect((0,0), (1, 2), horizontal: false)
  connect((0,0), (0, 2), horizontal: false)
  connect((0,0), (-1, 2), horizontal: false)
  connect((0,0), (-2, 2), horizontal: false)
})
```

![image](https://github.com/cetz-package/cetz/assets/519002/0fe42743-a476-4d66-947e-fe6c4cd6e293)
